modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -90,253 +92,107 @@ interface WalkResult {
-              while (textNode && textNode.type === 3) {
-                text = textNode.value;
-                sibling = textNode.prev;
-
-                if (text.length === 0 || isAllWhiteSpaceRegExp.test(text)) {
-                  textNode.remove();
-                  textNode = sibling;
-                }
-
-                textNode = sibling;
-              }
-            }
-          }
-        }
-
-        // Trim start white space
-        // Removed due to: #5424
-        /* textNode = node.prev;
-        if (textNode && textNode.type === 3) {
-          text = textNode.value.replace(startWhiteSpaceRegExp, '');
-
-          if (text.length > 0)
-            textNode.value = text;
-          else
-            textNode.remove();
-        }*/
+    }
+    if (rule.removeEmptyAttrs && ele.attributes.length === 0) {
+      Remove.unwrap(SugarElement.fromDom(ele));
+      return;
+    }
+  });
+
+  // Let's do the same thing for attributes
+  purify.addHook('uponSanitizeAttribute', (ele, evt) => {
+    evt.keepAttr = evt.attrName.startsWith('data-') || schema.isValid(ele.tagName.toLowerCase(), evt.attrName);
+    if (evt.keepAttr) {
+      if (!Obj.has(evt.allowedAttributes, evt.attrName)) {
+        evt.allowedAttributes[evt.attrName] = true;
       }
 
-      // Check if we exited a whitespace preserved element
-      if (isInWhiteSpacePreservedElement && whiteSpaceElements[name]) {
-        isInWhiteSpacePreservedElement = false;
+      if (evt.attrName in schema.getBoolAttrs()) {
+        evt.attrValue = evt.attrName;
       }
+    }
+  });
 
-      if (elementRule.removeEmpty && isEmpty(schema, nonEmptyElements, whiteSpaceElements, node)) {
-        tempNode = node.parent;
-
-        if (blockElements[node.name]) {
-          node.empty().remove();
-        } else {
-          node.unwrap();
-        }
+  purify.setConfig(config);
 
-        node = tempNode;
-        return;
-      }
+  return purify;
+};
 
-      if (elementRule.paddEmpty && (isPaddedWithNbsp(node) || isEmpty(schema, nonEmptyElements, whiteSpaceElements, node))) {
-        paddEmptyNode(settings, args, blockElements, node);
+const transferChildren = (parent: AstNode, nativeParent: Node) => {
+  const isSpecial = Arr.contains([ 'script', 'style' ], parent.name);
+  Arr.each(nativeParent.childNodes, (nativeChild) => {
+    const child = new AstNode(nativeChild.nodeName.toLowerCase(), nativeChild.nodeType);
+
+    if (NodeType.isElement(nativeChild)) {
+      Arr.each(nativeChild.attributes, (attr) => {
+        child.attr(attr.name, attr.value);
+      });
+    } else if (NodeType.isComment(nativeChild) || NodeType.isText(nativeChild) || NodeType.isCData(nativeChild) || NodeType.isPi(nativeChild)) {
+      child.value = nativeChild.data;
+      if (isSpecial) {
+        child.raw = true;
       }
-
-      node = node.parent;
+    } else {
+      // TODO: figure out if anything can arrive in this branch, before merging
```

Comment:
Do we have a Jira for this or does the comment mean before merging this PR?

Reply:
- [ ] It meant before merging the PR. Need to figure this out